File: web/src/engine/domain/session.ts
// Portage de Domain/Session.swift.
import type { Sport, SessionIntent, TrainingPhase } from './types';

export type StepKind = 'warmup' | 'work' | 'recovery' | 'rest' | 'cooldown' | 'repeatBlock';

/** Cible d'un pas de séance (union discriminée = enum Swift à valeurs associées). */
export type StepTarget =
  | { kind: 'hrZone'; zone: number }
  | { kind: 'paceRange'; lowSecPerKm: number; highSecPerKm: number }
  | { kind: 'swimPaceRange'; lowSecPer100m: number; highSecPer100m: number }
  | { kind: 'powerRange'; lowW: number; highW: number }
  | { kind: 'rpe'; value: number }
  | { kind: 'free' };

export const Target = {
  hrZone: (zone: number): StepTarget => ({ kind: 'hrZone', zone }),
  paceRange: (lowSecPerKm: number, highSecPerKm: number): StepTarget => ({ kind: 'paceRange', lowSecPerKm, highSecPerKm }),
  swimPaceRange: (lowSecPer100m: number, highSecPer100m: number): StepTarget => ({ kind: 'swimPaceRange', lowSecPer100m, highSecPer100m }),
  powerRange: (lowW: number, highW: number): StepTarget => ({ kind: 'powerRange', lowW, highW }),
  rpe: (value: number): StepTarget => ({ kind: 'rpe', value }),
  free: (): StepTarget => ({ kind: 'free' }),
};

/** Durée d'un pas : temps, distance, ou longueurs de bassin. */
export type StepDuration =
  | { kind: 'time'; seconds: number }
  | { kind: 'distance'; meters: number }
  | { kind: 'lengths'; count: number; poolMeters: number };

export const Duration = {
  time: (seconds: number): StepDuration => ({ kind: 'time', seconds }),
  distance: (meters: number): StepDuration => ({ kind: 'distance', meters }),
  lengths: (count: number, poolMeters: number): StepDuration => ({ kind: 'lengths', count, poolMeters }),
};

/** Durée estimée en secondes, pour le calcul de charge et l'agenda. */
export function estimatedSeconds(d: StepDuration, paceSecPerKm: number | null): number {
  switch (d.kind) {
    case 'time':
      return d.seconds;
    case 'distance': {
      if (paceSecPerKm == null) return d.meters / 3.0; // défaut ~3 m/s
      return (d.meters / 1000.0) * paceSecPerKm;
    }
    case 'lengths': {
      const m = d.count * d.poolMeters;
      const pace = paceSecPerKm ?? 95.0 * 10; // ~1:35/100m → s/km équiv.
      return (m / 1000.0) * pace;
    }
  }
}

export interface WorkoutStep {
  kind: StepKind;
  duration: StepDuration;
  target: StepTarget;
  cue?: string;
  repeats?: number; // pour repeatBlock
  children?: WorkoutStep[];
}

export interface PlannedSession {
  id: string;
  date: Date;
  sport: Sport;
  intent: SessionIntent;
  title: string;
  steps: WorkoutStep[];
  estimatedLoad: number; // charge (TSS-like)
  estimatedDuration: number; // secondes
  notes: string;
  phase?: TrainingPhase;
}

let sessionCounter = 0;
export function newSessionId(): string {
  return `sess-${++sessionCounter}-${Math.random().toString(36).slice(2, 8)}`;
}

export function makePlannedSession(
  p: Omit<Partial<PlannedSession>, 'date' | 'sport' | 'intent' | 'title'> & {
    date: Date;
    sport: Sport;
    intent: SessionIntent;
    title: string;
  },
): PlannedSession {
  return {
    id: p.id ?? newSessionId(),
    steps: [],
    estimatedLoad: 0,
    estimatedDuration: 0,
    notes: '',
    ...p,
  };
}

File: Packages/TriathlonEngine/Sources/TriathlonEngine/Domain/Session.swift
import Foundation

public enum StepKind: String, Codable, Sendable {
    case warmup, work, recovery, rest, cooldown, repeatBlock
}

/// Cible d'un pas de séance.
public enum StepTarget: Codable, Sendable, Equatable {
    case hrZone(Int)
    case paceRange(lowSecPerKm: Double, highSecPerKm: Double)
    case swimPaceRange(lowSecPer100m: Double, highSecPer100m: Double)
    case powerRange(lowW: Double, highW: Double)
    case rpe(Int)                 // 1–10
    case free
}

/// Durée d'un pas : temps, distance, ou longueurs de bassin.
public enum StepDuration: Codable, Sendable, Equatable {
    case time(seconds: TimeInterval)
    case distance(meters: Double)
    case lengths(count: Int, poolMeters: Double)

    /// Durée estimée en secondes, pour le calcul de charge et l'agenda.
    public func estimatedSeconds(paceSecPerKm: Double?) -> TimeInterval {
        switch self {
        case .time(let s): return s
        case .distance(let m):
            guard let pace = paceSecPerKm else { return m / 3.0 } // défaut ~3 m/s
            return (m / 1000.0) * pace
        case .lengths(let count, let poolMeters):
            let m = Double(count) * poolMeters
            let pace = paceSecPerKm ?? (95.0 * 10) // ~1:35/100m → s/km équiv.
            return (m / 1000.0) * pace
        }
    }
}

public struct WorkoutStep: Codable, Sendable, Equatable {
    public var kind: StepKind
    public var duration: StepDuration
    public var target: StepTarget
    public var cue: String?
    public var repeats: Int?          // pour repeatBlock
    public var children: [WorkoutStep]?

    public init(
        kind: StepKind,
        duration: StepDuration,
        target: StepTarget = .free,
        cue: String? = nil,
        repeats: Int? = nil,
        children: [WorkoutStep]? = nil
    ) {
        self.kind = kind
        self.duration = duration
        self.target = target
        self.cue = cue
        self.repeats = repeats
        self.children = children
    }
}

public struct PlannedSession: Codable, Sendable, Equatable, Identifiable {
    public var id: UUID
    public var date: Date
    public var sport: Sport
    public var intent: SessionIntent
    public var title: String
    public var steps: [WorkoutStep]
    public var estimatedLoad: Double        // charge (TSS-like)
    public var estimatedDuration: TimeInterval
    public var notes: String
    public var phase: TrainingPhase?

    public init(
        id: UUID = UUID(),
        date: Date,
        sport: Sport,
        intent: SessionIntent,
        title: String,
        steps: [WorkoutStep] = [],
        estimatedLoad: Double = 0,
        estimatedDuration: TimeInterval = 0,
        notes: String = "",
        phase: TrainingPhase? = nil
    ) {
        self.id = id
        self.date = date
        self.sport = sport
        self.intent = intent
        self.title = title
        self.steps = steps
        self.estimatedLoad = estimatedLoad
        self.estimatedDuration = estimatedDuration
        self.notes = notes
        self.phase = phase
    }
}
